File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.py
class Solution(object):
    def letterCombinations(self, digits):
        """
        :type digits: str
        :rtype: List[str]
        """
        result = []
        
        data = dict()
        data['2'] = ["a","b","c"]
        data['3'] = ["d","e","f"]
        data['4'] = ['g','h','i']
        data['5'] = ['j','k','l']
        data['6'] = ['m','n','o']
        data['7'] = ['p','q','r','s']
        data['8'] = ['t','u','v']
        data['9'] = ['w','x','y','z']
        if not digits:
            return ""
        def adder(stringer, index):
            for d in data[digits[index]]:
                copy = str(stringer) + d
                copyindex = index + 1
                
                if copyindex > len(digits) - 1:
                    result.append(copy)
                else:
                    adder(copy, copyindex)
        adder("",0)
        return result
                

File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.java
class Solution {
    List<String> answer = new ArrayList<>();
    Map<Character,String> key = new HashMap<>();
    public List<String> letterCombinations(String digits) {
        if(digits.length() < 1){
            return answer;
        }
        key.put('2', "abc");
        key.put('3', "def");
        key.put('4', "ghi");
        key.put('5',"jkl");
        key.put('6',"mno");
        key.put('7',"pqrs");
        key.put('8',"tuv");
        key.put('9',"wxyz");
        thro("", 0, digits);
        return answer;
    }
    
    private void thro(String current, int index, String digits){
        if(index >= digits.length()){
            answer.add(current);
        }
        else{
            char curr = digits.charAt(index);
            String s = key.get(curr);
            for(int i = 0; i < s.length(); i++){
                String copy = current;
                copy+= s.charAt(i);
                thro(copy, index+1, digits);
            }
        }
        
        
    }
}